### PR TITLE
Disable query pushdown if the corresponding flag is set

### LINKED ISSUE #1912: Add a flag for disabling query pushdown
### What should be done?

Out compatibility tests ensure that pushdown works correctly, but we can't test _everything_. Eventually, our users or we ourselves will find a problem in our Go filtering code or pushed down queries, and it would not be easy to differentiate between the two. So we should introduce a way to diagnose those problems easier.

Let's add `--test-disable-pushdown` flag for that. When it is `true`, we should ignore `pgdb.SQLParam.Filter` and `tigrisdb.FetchParam.Filter` fields, fetch all documents, and filter them only on the Go side. (While we are on it, let's refactor `pgdb` and `tigrisdb` to make them more similar to each other; for example, it seems like `SQLParam` and `FetchParam` should not be called that differently.)

We should also add this flag to integration testing. When true, all pushdowns should be disabled, and tests that check them should be updated.

Let's not add a test configuration (Taskfile target, CI job) for that yet, but it should be possible – all tests should pass with and without that flag.

### Where?

All handlers, all integration tests, `pgdb` / `tigrisdb` unit tests.

### Definition of Done

- all handlers updated;
- unit tests added/updated;
- integration/compatibility tests added/updated;
- spot refactorings done;
- user documentation for flags updated.

### PR DESCRIPTION
**Pull request by rumyantseva**
# Description

Closes #1912.

# In this PR

Implementation:
- disable query pushdown for all the cases where queries are in use;
- add user documentaiton.

Spot refactoring in addition to the implementation:
- `pgdb.QueryParam` -> `pgdb.QueryParams`;
- `tigrisdb.QueryParam` -> `tigrisdb.QueryParams`.
- simplified `msg_delete` (for both `pg` and `tigris` handlers)

## Readiness checklist

* [ ] I added/updated unit tests.
* [ ] I added/updated integration/compatibility tests.
* [ ] I added/updated comments and checked rendering.
* [x] I made spot refactorings.
* [x] I updated user documentation.
* [ ] I ran `task all`, and it passed.
* [x] I ensured that PR title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

File: cmd/ferretdb/main.go [hunk 1]
```diff
@@ -201,7 +201,11 @@ func setupLogger(stateProvider *state.Provider) *zap.Logger {
 
 	// Similarly to Prometheus, unless requested, don't add UUID to all messages, but log it once at startup.
 	if !cli.Log.UUID {
-		startupFields = append(startupFields, zap.String("uuid", logUUID))
+		startupFields = append(
+			startupFields,
+			zap.String("uuid", logUUID),
+			zap.Bool("disable-pushdown", cli.Test.DisablePushdown),
```
Comment (AlekSi) [inline cmd/ferretdb/main.go:207]: It should not be under that `if`

File: internal/handlers/tigris/tigrisdb/query.go [hunk 2]
```diff
@@ -30,16 +30,15 @@ import (
...
+// QueryParams represents options/parameters used by the fetch/query.
+type QueryParams struct {
 	Filter     *types.Document
 	DB         string
 	Collection string
 }
 
 // QueryDocuments fetches documents from the given collection.
-func (tdb *TigrisDB) QueryDocuments(ctx context.Context, param *QueryParam) (iterator.Interface[int, *types.Document], error) {
+func (tdb *TigrisDB) QueryDocuments(ctx context.Context, param *QueryParams) (iterator.Interface[int, *types.Document], error) {
```
Comment (AlekSi) [inline internal/handlers/tigris/tigrisdb/query.go:41]:
```suggestion
func (tdb *TigrisDB) QueryDocuments(ctx context.Context, qp *QueryParams) (iterator.Interface[int, *types.Document], error) {
```

File: cmd/ferretdb/main.go (hunk 1 above)
Comment (rumyantseva) [reply cmd/ferretdb/main.go:207]: Ah! 😅

File: internal/handlers/common/filter.go [hunk 3]
```diff
@@ -55,6 +55,11 @@ func FilterDocument(doc, filter *types.Document) (bool, error) {
 
 // filterDocumentPair handles a single filter element key/value pair {filterKey: filterValue}.
 func filterDocumentPair(doc *types.Document, filterKey string, filterValue any) (bool, error) {
+	if strings.HasPrefix(filterKey, "$") {
```
Comment (rumyantseva) [inline internal/handlers/common/filter.go:58]: This check is moved to the top of this function to handle cases like `path.$something`. Without this change, `-disable-pushdown=false` and `-disable-pushdown=true` would give different results (cc @chilagrow, it's relevant to your task a bit).

File: internal/handlers/tigris/tigrisdb/query.go (hunk 2 above)
Comment (rumyantseva) [reply internal/handlers/tigris/tigrisdb/query.go]: fixed

File: cmd/ferretdb/main.go (hunk 1 above)
Comment (rumyantseva) [reply cmd/ferretdb/main.go]: fixed

File: website/docs/flags.md [hunk 4]
```diff
@@ -41,6 +41,8 @@ Some default values are overridden in [our Docker image](quickstart_guide/docker
 | `proxy-addr`           | See [Operation modes/Proxy](operation_modes.md#proxy)                     | `FERRETDB_PROXY_ADDR`           |                                         |
 | **Telemetry**          |                                                                           |                                 |                                         |
 | `telemetry`            | See [Configure telemetry](telemetry.md#configure-telemetry)               | `FERRETDB_TELEMETRY`            | `undecided`                             |
+| **Testing**            |                                                                           |                                 |                                         |
+| `test-disable-pushdown` | Disable pushing down queries to the backend (to only filter on FerretDB)  | `FERRETDB_TEST_DISABLE_PUSHDOWN` | `false`                                 |
```
Comment (rumyantseva) [inline website/docs/flags.md:45]: @AlekSi please suggest a better description (if any)

Comment (AlekSi) [reply website/docs/flags.md:45]: That's fine for now

File: cmd/ferretdb/main.go [hunk 5]
```diff
@@ -196,6 +196,7 @@ func setupLogger(stateProvider *state.Provider) *zap.Logger {
 		zap.String("package", info.Package),
 		zap.Bool("debugBuild", info.DebugBuild),
 		zap.Any("buildEnvironment", info.BuildEnvironment.Map()),
+		zap.Bool("disable-pushdown", cli.Test.DisablePushdown),
```
Comment (AlekSi) [inline cmd/ferretdb/main.go:199]:
```suggestion
		zap.Bool("disablePushdown", cli.Test.DisablePushdown),
```

Note that it uses camelCase

File: internal/handlers/pg/msg_findandmodify.go [hunk 6]
```diff
@@ -69,7 +69,7 @@ func (h *Handler) MsgFindAndModify(ctx context.Context, msg *wire.OpMsg) (*wire.
 		ctx = ctxWithTimeout
 	}
 
-	queryParam := pgdb.QueryParam{
+	queryParam := pgdb.QueryParams{
```
Comment (AlekSi) [inline internal/handlers/pg/msg_findandmodify.go:72]:
```suggestion
	queryParams := pgdb.QueryParams{
```

File: internal/handlers/pg/msg_findandmodify.go [hunk 7]
```diff
@@ -214,7 +214,7 @@ func (h *Handler) MsgFindAndModify(ctx context.Context, msg *wire.OpMsg) (*wire.
 type upsertParams struct {
 	hasUpdateOperators bool
 	query, update      *types.Document
-	queryParam         *pgdb.QueryParam
+	queryParam         *pgdb.QueryParams
```
Comment (AlekSi) [inline internal/handlers/pg/msg_findandmodify.go:217]:
```suggestion
	queryParams         *pgdb.QueryParams
```

File: internal/handlers/tigris/msg_find.go [hunk 8]
```diff
@@ -95,8 +95,16 @@ func (h *Handler) MsgFind(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, er
 	return &reply, nil
 }
 
-// fetchAndFilterDocs fetches documents from the database and filters them using the provided QueryParam.Filter.
-func fetchAndFilterDocs(ctx context.Context, dbPool *tigrisdb.TigrisDB, qp *tigrisdb.QueryParam) ([]*types.Document, error) {
+// fetchAndFilterDocs fetches documents from the database and filters them using the provided QueryParams.Filter.
+func (h *Handler) fetchAndFilterDocs(ctx context.Context, dbPool *tigrisdb.TigrisDB, qp *tigrisdb.QueryParams) ([]*types.Document, error) { //nolint:lll // argument list is too long
```
Comment (AlekSi) [inline internal/handlers/tigris/msg_find.go:99]: The problem with converting those functions back to methods (as they were before) is that receiver has a lot of extra information, including more `*tigrisdb.TigrisDB` connection pools. I think we should move to functions, not into the opposite direction

Comment (rumyantseva) [reply internal/handlers/tigris/msg_find.go:99]: @AlekSi so then we pass whatever we need from `Handler` as params? Do we need to introduce a new struct to pass params, or should we just extend the list of params?

Comment (AlekSi) [reply internal/handlers/tigris/msg_find.go:99]: Up to you

**Review by w84thesun — COMMENTED** [review]
Overall looks good, not much to add to @AlekSi's comments.

File: internal/handlers/pg/msg_findandmodify.go (hunk 6 above)
Comment (rumyantseva) [reply internal/handlers/pg/msg_findandmodify.go]: fixed

File: internal/handlers/pg/msg_findandmodify.go (hunk 7 above)
Comment (rumyantseva) [reply internal/handlers/pg/msg_findandmodify.go]: fixed

File: internal/handlers/tigris/msg_find.go (hunk 8 above)
Comment (rumyantseva) [reply internal/handlers/tigris/msg_find.go]: fixed

File: internal/handlers/pg/msg_delete.go [hunk 9]
```diff
@@ -180,8 +180,17 @@ func (h *Handler) prepareDeleteParams(deleteDoc *types.Document) (*types.Documen
 // execDelete fetches documents, filters them out, limits them (if needed) and deletes them.
 // If limit is true, only the first matched document is chosen for deletion, otherwise all matched documents are chosen.
 // It returns the number of deleted documents or an error.
-func execDelete(ctx context.Context, dbPool *pgdb.Pool, qp *pgdb.QueryParam, limit bool) (int32, error) {
+func (h *Handler) execDelete(ctx context.Context, dbPool *pgdb.Pool, qp *pgdb.QueryParams, limit bool) (int32, error) {
```
Comment (w84thesun) [inline internal/handlers/pg/msg_delete.go:183]: Should we remove the receiver here too?

Comment (rumyantseva) [reply internal/handlers/pg/msg_delete.go:183]: Yeah, sounds reasonable!

File: internal/handlers/common/filter.go (hunk 3 above)
Comment (chilagrow) [reply internal/handlers/common/filter.go:58]: Thanks!

File: internal/handlers/pg/msg_find.go [hunk 10]
```diff
@@ -110,9 +110,24 @@ func (h *Handler) MsgFind(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, er
...
-func (h *Handler) fetchAndFilterDocs(ctx context.Context, tx pgx.Tx, sqlParam *pgdb.QueryParam) ([]*types.Document, error) {
-	iter, err := pgdb.QueryDocuments(ctx, tx, sqlParam)
+func fetchAndFilterDocs(ctx context.Context, fp *fetchParams) ([]*types.Document, error) {
+	// filter is used to filter documents on the FerretDB side,
+	// qp.Filter is used to filter documents on the PostgreSQL side (query pushdown).
+	filter := fp.qp.Filter
+
+	if fp.disablePushdown {
+		fp.qp.Filter = nil
+	}
```
Comment (chilagrow) [inline internal/handlers/pg/msg_find.go:128]: Do you need a unit test for this and tigris? Integration test isn't in scope as I read the issue.

**Review by chilagrow — DISMISSED** [review]

File: internal/handlers/pg/msg_find.go (hunk 10 above)
Comment (rumyantseva) [reply internal/handlers/pg/msg_find.go:128]: I believe not for now. When we set `disable-pushdown` flag with running tests, there will be checks (implemented in a separate PR).

**Review by Fashander — DISMISSED** [review]

File: internal/handlers/pg/msg_delete.go (hunk 9 above)
Comment (rumyantseva) [reply internal/handlers/pg/msg_delete.go]: fixed

**Review by w84thesun — DISMISSED** [review]
👍

**Review by chilagrow — DISMISSED** [review]
Great work!

File: internal/handlers/common/filter.go (hunk 3 above)
Comment (AlekSi) [reply internal/handlers/common/filter.go:58]: That is actually quite concerning: `path.$something` should return the same result with pushdown enabled or disabled – an error about invalid path / field name. And I think we should a missing check for in the right place and add tests, not just move `if`

**Review by AlekSi — COMMENTED** [review]
https://github.com/FerretDB/FerretDB/pull/2003/files#r1113844095

**Review by Fashander — DISMISSED** [review]

File: internal/handlers/common/filter.go (hunk 3 above)
Comment (rumyantseva) [reply internal/handlers/common/filter.go:58]: Interesting... I moved this back, and I can't find this case anymore. So, maybe it was something else. 

Then I don't introduce this change, but I added a couple of more tests with `$`.

**Review by Fashander — DISMISSED** [review]

**Review by w84thesun — APPROVED** [review]

File: internal/handlers/pg/msg_delete.go [hunk 11]
```diff
@@ -167,23 +161,40 @@ func (h *Handler) prepareDeleteParams(deleteDoc *types.Document) (*types.Documen
...
 			"limit",
 		)
 	}
 
-	return filter, limit, nil
+	return filter, limit == 1, nil
+}
+
+// deleteParams contains parameters for execDelete function.
+type deleteParams struct {
```
Comment (noisersup) [inline internal/handlers/pg/msg_delete.go:175]: I really like idea of creating parameter structs for those functions, although the name here seems misleading to me. Now we have `deleteParams` struct which is used to provide arguments to `execDelete` function, but in the same file we also provide `prepareDeleteParams` function which extracts parameter for mongo query. Can we rename it to let's say `execDeleteParams`?

**Review by noisersup — CHANGES_REQUESTED** [review]
Looks great! Setting filter to nil on handler level seems much better! One small comment about struct naming.

**Review by w84thesun — APPROVED** [review]
👍

File: internal/handlers/pg/msg_delete.go (hunk 11 above)
Comment (rumyantseva) [reply internal/handlers/pg/msg_delete.go]: Fixed, please take another look

**Review by noisersup — APPROVED** [review]
:shipit:

**Review by Fashander — APPROVED** [review]

**Pull request merged**